### PR TITLE
Add Cross-domain Watcher service

### PR DESCRIPTION
**Pull request by K-Ho**
## Description
This adds a cross domain watcher service. Usage is as follows: 
Watcher is initialized with L1 and L2 providers + L1 and L2 messenger addresses.

User sends an L1->L2 Deposit tx. Front-end gets the L1 tx hash and calls `getMessageHashesFromL1Tx` which returns an array of all messageHashes of the L1->L2 txs that were sent in this tx (we would expect this to be a singleton array, but you could construct an L1 tx that sends multiple L1->L2 messages).

Front-end now calls `onceL2Relay(messageHash, (l2TxHash) => {// do thing})`. Once the L2 tx is relayed, then the callback will be triggered with the L2 tx hash.

The watcher service can of course be used in the reverse direction as well, using `getMessageHashesFromL2Tx` followed by `onceL1Relay(messageHash, (l1TxHash) => {// do thing})`

## Questions
Best way to add tests for this other than the integration testing repo?

### Fixes
- Fixes #https://github.com/ethereum-optimism/roadmap/issues/54

## Contributing Agreement

- [X] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

**Review by tynes — APPROVED** [review]

File: packages/ovm-toolchain/src/watcher.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  public async getMessageHashesFromL1Tx(l1TxHash: string): Promise<string[]> {
+    return this._getMessageHashesFromTx(true, l1TxHash)
+  }
+  public async getMessageHashesFromL2Tx(l2TxHash: string): Promise<string[]> {
+    return this._getMessageHashesFromTx(false, l2TxHash)
+  }
+
+  public onceL2Relay(msgHash: string, callback: Function) {
```
Comment (kfichter) [inline packages/ovm-toolchain/src/watcher.ts:30]: Nice, I like this API. LGTM :+1:

**Review by kfichter — APPROVED** [review]
API looks good!

File: packages/ovm-toolchain/src/watcher.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    return this._onceRelay(false, msgHash, callback)
+  }
+
+  public onceL1Relay(msgHash: string, callback: Function) {
```
Comment (karlfloersch) [inline packages/ovm-toolchain/src/watcher.ts:34]: What were the thoughts here about using a callback vs just returning a promise?

Comment (karlfloersch) [reply packages/ovm-toolchain/src/watcher.ts:34]: Oh also should probably add a return value for this function

Comment (K-Ho) [reply packages/ovm-toolchain/src/watcher.ts:34]: Was trying to emulate the `provider.once` functionality in `ethers` https://docs.ethers.io/v5/single-page/#/v5/api/providers/provider/-%23-Provider-once -

Comment (tynes) [reply packages/ovm-toolchain/src/watcher.ts:34]: Another way to do this that is a bit more flexible I think is to make the `Watcher` an `EventEmitter` and then update this part of the code:

```js
    layer.provider.on(filter, (log: any) => {
      if (log.data === msgHash) {
        this.emit('relayed', log.transactionHash)
      }
```

Then the `Watcher` would have the `on` property and the `once` property as well as the rest of the highly used/tested `EventEmitter` properties

**Review by karlfloersch — APPROVED** [review]
LGTM! 🚢🛠🚀

**Pull request merged**